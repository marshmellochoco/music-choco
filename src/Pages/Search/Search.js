// dependancy import
import axios from "axios";
import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";

// component import
import { Card } from "../../Components/Card/Card";
import { addQueue } from "../../store/actions/queueAction";
import { setPlayingSong } from "../../store/actions/songDataAction";
import "./Search.css";

export const Search = ({ search }) => {
    const apiUrl = process.env.REACT_APP_API_URL;
    const [albumSearch, setAlbumSearch] = useState([]);
    const [songSearch, setSongSearch] = useState([]);

    const playingSong = useSelector(
        (state) => state.songDataReducer.songData.songId
    );

    const dispatch = useDispatch();

    useEffect(() => {
        if (search && search !== "") {
            axios
                .get(`${apiUrl}/album/search/${search}`)
                .then((result) => setAlbumSearch(result.data))
                .catch((e) => setAlbumSearch([]));

            axios
                .get(`${apiUrl}/song/search/${search}`)
                .then((result) => setSongSearch(result.data))
                .catch((e) => setSongSearch([]));
        }
    }, [search, apiUrl]);

    const playSong = (id) => {
        dispatch(addQueue(id));
        dispatch(setPlayingSong(id));
        dispatch({ type: "SET_PLAYING", playing: true });
    };

    const addSong = (id) => {
        // add the song to the queue
        dispatch(addQueue(id));
    };

    const handleImageError = (e) => {
        // return an alternate image if error
        e.target.src = "https://f4.bcbits.com/img/a4139357031_10.jpg";
    };

    const searchAlbum = () => {
        // return a list of card item of the album search result as component list
        return albumSearch.map((s) => {
            return (
                <Card
                    key={s._id}
                    id={s._id}
                    name={s.albumname}
                    artist={s.artist}
                    handleImageError={handleImageError}
                ></Card>
            );
        });
    };

    const searchSong = () => {
        // return a list of item of the song title search result as component list
        return songSearch.map((s) => {
            return (
                <li
                    key={s.songs._id}
                    onClick={(e) => playSong(s.songs._id)}
                    onContextMenu={(e) => addSong(s.songs._id)}
                    className={playingSong === s.songs._id ? "activeItem" : ""}
                >
                    <img
                        src={`${apiUrl}/album/ico/${s._id}`}
                        alt="album logo"
                        className="album"
                        onError={handleImageError}
                    />
                    <div className="songDetail">
                        <div className="songTitle">{s.songs.title}</div>
                        <div>{s.artist}</div>
                    </div>
                    <div>
                        {(s.songs.duration / 60 < 10 ? "0" : "") +
                            Math.floor(s.songs.duration / 60)}{" "}
                        :{" "}
                        {(s.songs.duration % 60 < 10 ? "0" : "") +
                            (s.songs.duration % 60)}
                    </div>
                </li>
            );
        });
    };

    return (
        <div>
            {search === "" ? (
                <div>Hello</div>
            ) : (
                <div className="content">
                    <h1>Albums</h1>
                    <div className="cardList">{searchAlbum()}</div>
                    <h1>Songs</h1>
                    <ul className="songList">{searchSong()}</ul>
                </div>
            )}
        </div>
    );
};
