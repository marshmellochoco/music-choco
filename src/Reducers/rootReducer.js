import { queueReducer } from "../Reducers/queueReducer";
import { songDataReducer } from "../Reducers/songDataReducer";
import { playerReducer } from "../Reducers/playerReducer";
import { authReducer } from "../Reducers/authReducer";

import { combineReducers } from "redux";

export const rootReducer = combineReducers({
    queueReducer,
    songDataReducer,
    playerReducer,
    authReducer,
});
