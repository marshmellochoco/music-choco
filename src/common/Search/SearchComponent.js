/** @jsxImportSource @emotion/react */
import { css } from "@emotion/react";
import Icon from "@mdi/react";
import { mdiPlayCircle } from "@mdi/js";
import { Link } from "react-router-dom";

export const SearchComponent = ({
    imageUrl,
    playingSong,
    albumSearch,
    songSearch,
    playSong,
    addSong,
    handleImageError,
    handlePlayAlbum,
}) => {
    // styles
    const albumListStyle = css`
        display: flex;
        overflow-x: scroll;

        & > a {
            color: white;
            text-decoration: none;
            &:hover {
                text-decoration: underline;
            }
        }

        & .cardItem {
            margin-right: 1rem;
            margin-bottom: 1rem;
        }
    `;

    const albumCardStyle = css`
        width: 200px;
        height: 300px;
        padding: 1rem;
        margin-right: 2rem;
        border-radius: 2%;
        color: var(--text-color);
        background-color: var(--secondary-color);
        cursor: pointer;

        & > img {
            width: 200px;
        }
        & > div {
            display: flex;
            flex-direction: column;
            & > h1 {
                height: 3em;
                overflow: hidden;
                text-overflow: ellipsis;
                font-size: 1.4em;
                margin: 0;
            }
            & > div {
                display: flex;
                justify-content: space-between;
                & > p {
                    overflow: hidden;
                    font-size: 1.2em;
                    margin: 0;
                }
                & > svg {
                    width: 2rem;
                    &:hover {
                        opacity: 0.7;
                    }
                }
            }
        }
    `;

    const songListStyle = css`
        list-style-type: none;
        padding: 0;
        margin-right: 1rem;

        & > li {
            padding: 0.5rem 1rem;
            height: 60px;
            display: grid;
            grid-template-columns: 70px 20fr 4em;
            cursor: pointer;

            &:hover {
                background-color: var(--secondary-color);
            }

            & > div {
                display: flex;
                align-items: center;
            }

            & > .songDetail {
                display: grid;
                grid-template-rows: 1fr 1fr;
            }

            & > .songTitle {
                font-size: 1.2em;
                overflow: hidden;
            }

            & > .album {
                width: 60px;
            }
        }
        & .activeItem {
            background-color: var(--thirtiary-color);
            color: var(--primary-color);
        }
    `;

    // markdown
    const albumList = (albums) => {
        return albums.map((s) => {
            return (
                <Link css={albumCardStyle} to={`/albums/${s._id}`} key={s._id}>
                    <img
                        src={`${imageUrl}${s._id}`}
                        alt="Album Icon"
                        onError={handleImageError}
                    />
                    <div>
                        <h1>{s.albumname}</h1>
                        <div>
                            <p>{s.artist}</p>
                            <Icon
                                path={mdiPlayCircle}
                                onClick={(e) => {
                                    e.preventDefault();
                                    handlePlayAlbum(s._id);
                                }}
                            />
                        </div>
                    </div>
                </Link>
            );
        });
    };

    const songList = (songs) => {
        return songs.map((s) => {
            return (
                <li
                    key={s.songs._id}
                    onClick={(e) => playSong(s.songs._id)}
                    onContextMenu={(e) => addSong(s.songs._id)}
                    className={playingSong === s.songs._id ? "activeItem" : ""}
                >
                    <img
                        src={`${imageUrl}${s._id}`}
                        alt="album logo"
                        className="album"
                        onError={handleImageError}
                    />
                    <div className="songDetail">
                        <div className="songTitle">{s.songs.title}</div>
                        <div>{s.artist}</div>
                    </div>
                    <div>
                        {(s.songs.duration / 60 < 10 ? "0" : "") +
                            Math.floor(s.songs.duration / 60)}{" "}
                        :{" "}
                        {(s.songs.duration % 60 < 10 ? "0" : "") +
                            (s.songs.duration % 60)}
                    </div>
                </li>
            );
        });
    };

    return (
        <div>
            <h1>Albums</h1>
            <div css={albumListStyle}>{albumList(albumSearch)}</div>
            <h1>Songs</h1>
            <ul css={songListStyle}>{songList(songSearch)}</ul>
        </div>
    );
};
