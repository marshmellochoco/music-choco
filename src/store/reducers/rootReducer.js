import { queueReducer } from "./queueReducer";
import { songDataReducer } from "./songDataReducer";
import { playerReducer } from "./playerReducer";
import { authReducer } from "./authReducer";

import { combineReducers } from "redux";

export const rootReducer = combineReducers({
    queueReducer,
    songDataReducer,
    playerReducer,
    authReducer,
});
