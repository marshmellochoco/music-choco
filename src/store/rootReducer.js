import { combineReducers } from "redux";
import queueReducer from "./queue/queueReducer";
import playerReducer from "./player/playerReducer";
import userReducer from "./user/userReducer";

const rootReducer = combineReducers({
    playerReducer,
    queueReducer,
    userReducer,
});

export default rootReducer;
