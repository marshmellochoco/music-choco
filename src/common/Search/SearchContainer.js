import axios from "axios";
import { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { SearchComponent } from "./SearchComponent";
import { setPlayingSong } from "../../store/actions/songDataAction";
import { addQueue, setQueue } from "../../store/actions/queueAction";
import { setPlaying } from "../../store/actions/playerActions";

export const SearchContainer = ({ search }) => {
    const apiUrl = process.env.REACT_APP_API_URL;
    const imageUrl = `${apiUrl}/album/ico/`;
    const [albumSearch, setAlbumSearch] = useState([]);
    const [songSearch, setSongSearch] = useState([]);
    const playingSong = useSelector(
        (state) => state.songDataReducer.songData.songId
    );

    const dispatch = useDispatch();

    useEffect(() => {
        if (search && search !== "") {
            axios
                .get(`${apiUrl}/album/search/${search}`)
                .then((result) => setAlbumSearch(result.data))
                .catch((e) => setAlbumSearch([]));

            axios
                .get(`${apiUrl}/song/search/${search}`)
                .then((result) => setSongSearch(result.data))
                .catch((e) => setSongSearch([]));
        }
    }, [search, apiUrl]);

    const playSong = (id) => {
        dispatch(addQueue(id));
        dispatch(setPlayingSong(id));
        dispatch(setPlaying(true));
    };

    const addSong = (id) => {
        // add the song to the queue
        dispatch(addQueue(id));
    };

    const handleImageError = (e) => {
        // return an alternate image if error
        e.target.src = "https://f4.bcbits.com/img/a4139357031_10.jpg";
    };

    const handlePlayAlbum = async (id) => {
        // play the entire album associated with the card, replace the existing queue with the album list
        var songListData = [];
        await axios
            .get(`${apiUrl}/album/${id}`)
            .then((res) => (songListData = res.data.songs))
            .catch((e) => (songListData = []));

        var songList = [];
        songListData.forEach((s) => {
            songList.push(s._id);
        });

        dispatch(setPlayingSong(songList[0]));
        dispatch(setPlaying(true));
        dispatch(setQueue(songList));
    };

    return (
        <SearchComponent
            imageUrl={imageUrl}
            search={search}
            playingSong={playingSong}
            albumSearch={albumSearch}
            songSearch={songSearch}
            playSong={playSong}
            addSong={addSong}
            handleImageError={handleImageError}
            handlePlayAlbum={handlePlayAlbum}
        />
    );
};
