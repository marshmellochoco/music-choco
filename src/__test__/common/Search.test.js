import { configure, mount } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import toJson from "enzyme-to-json";
import { Provider } from "react-redux";
import { createStore } from "redux";

import { rootReducer } from "../../store/rootReducer";
import { SearchContainer } from "../../common/Search/SearchContainer";

configure({ adapter: new Adapter() });
const store = createStore(rootReducer);

describe("Search", () => {
    it("should render the search", () => {
        const wrapper = mount(
            <Provider store={store}>
                <SearchContainer />
            </Provider>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
