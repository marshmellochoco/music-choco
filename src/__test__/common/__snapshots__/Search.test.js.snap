// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render the search 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <SearchContainer>
    <SearchComponent
      addSong={[Function]}
      albumSearch={Array []}
      handleImageError={[Function]}
      handlePlayAlbum={[Function]}
      imageUrl="http://118.100.207.12:4000/api/album/ico/"
      playSong={[Function]}
      playingSong=""
      songSearch={Array []}
    >
      <div>
        <h1>
          Albums
        </h1>
        <EmotionCssPropInternal
          __EMOTION_LABEL_PLEASE_DO_NOT_USE__="SearchComponent"
          __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
          css={
            Object {
              "map": undefined,
              "name": "141xi7h",
              "next": undefined,
              "styles": "
        display: flex;
        overflow-x: scroll;

        & > a {
            color: white;
            text-decoration: none;
            &:hover {
                text-decoration: underline;
            }
        }

        & .cardItem {
            margin-right: 1rem;
            margin-bottom: 1rem;
        }
    ",
              "toString": [Function],
            }
          }
        >
          <div
            className="css-1zowee-SearchComponent"
          />
        </EmotionCssPropInternal>
        <h1>
          Songs
        </h1>
        <EmotionCssPropInternal
          __EMOTION_LABEL_PLEASE_DO_NOT_USE__="SearchComponent"
          __EMOTION_TYPE_PLEASE_DO_NOT_USE__="ul"
          css={
            Object {
              "map": undefined,
              "name": "h43zs4",
              "next": undefined,
              "styles": "
        list-style-type: none;
        padding: 0;
        margin-right: 1rem;

        & > li {
            padding: 0.5rem 1rem;
            height: 60px;
            display: grid;
            grid-template-columns: 70px 20fr 4em;
            cursor: pointer;

            &:hover {
                background-color: var(--secondary-color);
            }

            & > div {
                display: flex;
                align-items: center;
            }

            & > .songDetail {
                display: grid;
                grid-template-rows: 1fr 1fr;
            }

            & > .songTitle {
                font-size: 1.2em;
                overflow: hidden;
            }

            & > .album {
                width: 60px;
            }
        }
        & .activeItem {
            background-color: var(--thirtiary-color);
            color: var(--primary-color);
        }
    ",
              "toString": [Function],
            }
          }
        >
          <ul
            className="css-m6uak6-SearchComponent"
          />
        </EmotionCssPropInternal>
      </div>
    </SearchComponent>
  </SearchContainer>
</Provider>
`;
